groth16/src/prove.rs

```diff
@@ -1,119 +1,137 @@
-pub async fn compute_C<
-    E: Pairing<G1Affine = F>,
-    F: FftField + PrimeField,
-    Net: MpcNet,
->(
-    A: E::G1Affine,
-    M: E::G1,
-    s: F,
-    r: F,
-    (w_pp, W): (PackedSharingParams<E::ScalarField>, Vec<E::G1Affine>),
-    (u_pp, U): (PackedSharingParams<E::ScalarField>, Vec<E::G1Affine>),
-    (h_pp, H): (PackedSharingParams<E::ScalarField>, Vec<E::G1Affine>),
-    a: Vec<E::ScalarField>,
-    h: Vec<E::ScalarField>,
-    net: &Net,
-) -> Result<E::G1Affine, MpcNetError> {
-    // We use variables (A, M, ∏{i∈[l+1,m]}(W_i)^a_i, ∏{i∈[0,Q−2]}(U_i)h_i, ∏{i∈[0,m]}(H_i)^a_i)
-    // to denote elements in G1. We also assume that all the servers computing the proof
-    // get A, M, s, r and h in the clear and only receive packed shares of the remaining elements.
-
-    const CHANNEL0: MultiplexedStreamID = MultiplexedStreamID::Zero;
-    const CHANNEL1: MultiplexedStreamID = MultiplexedStreamID::One;
-    const CHANNEL2: MultiplexedStreamID = MultiplexedStreamID::Two;
-
-    // Calculate ∏{i∈[l+1,m]}(W_i)^a_i using dmsm
-    let w = d_msm::<E::G1, _>(&W, &a, &w_pp, net, CHANNEL0);
-    // Calculate ∏{i∈[0,Q−2]}(U_i)^h_i using dmsm
-    let u = d_msm::<E::G1, _>(&U, &h, &u_pp, net, CHANNEL1);
-    // Calculate ∏{i∈[0,m]}(H_i)^a_i using dmsm
-    let h = d_msm::<E::G1, _>(&H, &a, &h_pp, net, CHANNEL2);
-
-    let (w, u, h) = tokio::try_join!(w, u, h)?;
-
-    // Calculate A^s
-    let v0 = A.pow(s.into_bigint());
-    // Calculate M^r
-    let v1 = M.into_affine().pow(r.into_bigint());
-    // Calculate (∏{i∈[0,m]}(H_i)^a_i)^r
-    let v2 = h.into_affine().pow(r.into_bigint());
-    // finally compute C
-    let c = w.into_affine().mul(u.into_affine()).mul(v0).mul(v1).mul(v2);
-    Ok(c)
+#[derive(Debug, Clone, Copy)]
+pub struct C<'a, E: Pairing> {
+    pub A: E::G1,
+    pub M: E::G1Affine,
+    pub s: E::ScalarField,
+    pub r: E::ScalarField,
+    pub pp: &'a PackedSharingParams<E::ScalarField>,
+    pub W: &'a [E::G1Affine],
+    pub U: &'a [E::G1Affine],
+    pub H: &'a [E::G1Affine],
+    pub a: &'a [E::ScalarField],
+    pub ax: &'a [E::ScalarField],
+    pub h: &'a [E::ScalarField],
+}
+
+impl<'a, E: Pairing> C<'a, E> {
+    /// Computes C
+    pub async fn compute<Net: MpcNet>(
+        self,
+        net: &Net,
+    ) -> Result<E::G1, MpcNetError> {
+        // We use variables (A, M, ∏{i∈[l+1,m]}(W_i)^a_i, ∏{i∈[0,Q−2]}(U_i)h_i, ∏{i∈[0,m]}(H_i)^a_i)
+        // to denote elements in G1. We also assume that all the servers computing the proof
+        // get A, M, s, r and h in the clear and only receive packed shares of the remaining elements.
+
+        const CHANNEL0: MultiplexedStreamID = MultiplexedStreamID::Zero;
+        const CHANNEL1: MultiplexedStreamID = MultiplexedStreamID::One;
+        const CHANNEL2: MultiplexedStreamID = MultiplexedStreamID::Two;
+
+        // Calculate ∏{i∈[l+1,m]}(W_i)^a_i using dmsm
+        let w = d_msm::<E::G1, _>(self.W, self.ax, self.pp, net, CHANNEL0);
```

Comment:
Should these calls be `.await`ed?

Reply:
They are all awaited at the end, computing them concurrently.

Reply:
Ah I see!